nautobot/nautobot
# Handle time zones more correctly with ScheduledJob

# Closes #5591
# What's Changed

- Because Django stores UTC in all `DateTimeFields`, but we want to be able to have our Scheduled Jobs respect things like DST where applicable, potentially even including jobs in different time zones (e.g. 5 PM EDT, 5 PM PDT), add a `time_zone` field to the `ScheduledJob` model. As part of the migrations, all existing ScheduledJobs will gain `time_zone = settings.TIME_ZONE` as their default.
- Updated `ScheduledJob.to_cron` and `ScheduledJob.get_crontab` methods to make use of the `time_zone` and return a django-celery-beat `TzAwareCrontab` instead of a vanilla celery `crontab`.
- Add a bunch of test cases in `test_models` module to exercise and validate this behavior, making heavy use of the `time-machine` Python package as an added test dependency.

Additionally:

- ScheduledJob model had no default `ordering`, so I've added one for more consistent UI behavior.
- Added `time_zone` and `crontab` columns to `ScheduledJobTable`.
- Updated ScheduledJob detail view to display the `start_time` and `last_run_at` according to both the job's `time_zone` and UTC.
  - [x] ~TODO: There is probably a cleaner way to do this, perhaps with some additional `@property` on `ScheduledJob`, so that for example we don't show the time twice if the time_zone is the same in both cases?~
- Because I got tired of messing up my manual testing because I forgot to restart the `celery_beat` dev container after making code changes, I added the `watchmedo` pattern so that it auto-restarts when Nautobot code changes, just like the `celery_worker` container was already doing. 
- A few minor test updates to have more "realistic" ScheduledJob data, not directly necessary for this fix.

Manual Testing:

- With settings.TIME_ZONE set to `America/New_York`:
    - Create a ScheduledJob of interval DAILY via the UI and verify that its time_zone defaults to New York
    - Verify that it runs at the expected time
    - Create a ScheduledJob of interval CRONTAB via the UI and verify that its time_zone defaults to New York
    - Verify that it runs at the expected time
    - Create a ScheduledJob of interval DAILY via the UI, update it via nbshell to "America/Los_Angeles" time zone, and verify that it runs at the expected time
    - Create a ScheduledJob of interval CRONTAB via the UI, update it via nbshell to "America/Los_Angeles" time zone, and verify that it runs at the scheduled time
- [x] TODO: Change settings.TIME_ZONE to `UTC`, restart all containers, and verify that all jobs re-run at the expected times and not at +/- 4 hours as described in #5591.


![image](https://github.com/user-attachments/assets/3b9d7083-9443-47b8-b668-5ccb39eade2e)

![image](https://github.com/user-attachments/assets/d1362957-bf47-48f8-bf8b-8afac572e4b2)


# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Documentation Updates (when adding/changing features)
- n/a Example App Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design
